matrix-free: do not use bool args

tjhei opened the pull request:
Functions with several bool arguments in matrix-free have annoyed me for quite a while but seeing Martin show the code today, made me propose this.

This is obviously incomplete, but I am looking for comments now. I just started with:

- replace evaluate*, integrate* calls in FEEvaluation

Note: I am not using an ``enum class`` because that doesn't behave well with flags to ``or`` together. I put it in a namespace for scoping, but that is debatable.

bangerth commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -57,6 +57,86 @@ template <int dim,
+  /**
+   * @brief The type enum
+   */
+  enum type
```
Maybe call it `EvaluationFlags` as well? Also, the documentation above shows up in doxygen but doesn't give anything away if you miss the name of the surrounding namespace. I think you might want to say a couple more words there :-)

tjhei replied:
hm, that makes it ``EvaluationFlags::EvaluationFlags`` in the function args, which is a bit weird. Do we have an example somewhere in the library where we wrap an enum like this?

bangerth commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -7018,6 +7150,49 @@ FEEvaluation<dim,
+
+#  ifdef DEBUG
+  if (evaluation_flags & EvaluationFlags::values)
+    this->values_quad_initialized = true;
```
You might as well replace these three member variables with your enum as well :-)

kronbichler replied:
I can take those internal parts over.

kronbichler commented:
I also like this - I thought about a better interface back in 2011 long before this was merged into deal.II, but then nobody did pick things up again. Thanks @tjhei.
Now we only need to switch all programs to this notation. :wink:

kronbichler commented:
Let us know if you need help with this, because there are certainly a hundred or more of those functions in the testsuite alone. I would also pull it through all internal interfaces, but I can take over that part. Regarding the names, I agree with @bangerth.

drwells commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -57,6 +57,86 @@ template <int dim,
+    nothing   = 0,
+    values    = 0x1,
+    gradients = 0x2,
+    hessians  = 0x4
```
It's not a big deal, but could we name these `evaluate_default`, `evaluate_values`, etc. so that they agree with `UpdateFlags`?

kronbichler replied:
One thing I would like to avoid is long names that contain redundancies - we already have `EvaluationFlags::` in the name, so it feels superfluous. These names are typed quite often in a matrix-free code.

tjhei replied:
I liked the shorter notation because ``integrate(EvaluationFlags::evaluate_values);`` is verbose. Thoughts?

drwells replied:
oh, these have a namespace. Sure.

tjhei commented:
I updated to use ``EvaluationFlags::EvaluationFlags`` and updated the whole interface (I think). If we agree on this, I will start with the examples/tests.

kronbichler commented:
I think this looks good. The first thing that I think we should assess is the question

> that makes it `EvaluationFlags::EvaluationFlags` in the function args, which is a bit weird. Do we have an example somewhere in the library where we wrap an enum like this?

It would still be interesting to have a look and see what patterns we have for enums. My feeling says we're mixing many variants. The thing is, the more time I think about it the more I also see attraction in @tjhei's original variant - the argument type can be passed on to user code if one nests these evaluators a bit to support multiple equations via a common interface, so it is not just `EvaluationFlags::values` a user might write. But then again, the new interface will bring relief.

The second question is whether we deprecate the old interfaces straight away or let a little time pass to see the new variant in action. I think we all agree that the new interface is better, but before we deprecate and make application codes change to avoid the deprecation I would a little time pass. I admit I am biased because I am the person most strongly affected, but hesitate to do things before we are really convinced that we have it nice now.

bangerth commented:
On 5/27/20 12:49 PM, Martin Kronbichler wrote:
> that makes it |EvaluationFlags::EvaluationFlags| in the function args, which 
> is a bit weird. Do we have an example somewhere in the library where we wrap 
> an enum like this?

Yes: |DataComponentInterpretation::DataComponentInterpretation|

tjhei commented:
I just discovered DataAccessOnFaces [1] that is an enum class that has similar members. Thoughts on unifying them?

[1]: https://www.dealii.org/developer/doxygen/deal.II/classMatrixFree.html#a751b5f06534d52773eba384d2c866a15

tjhei commented:
I just grepped through the whole library and replaced all function calls.

That leaves the discussion about whether to deprecate now or in the future. I don't feel strongly about it.

kronbichler commented:
It definitely makes sense to unify them because they are targeting the same thing at the end of the day - the argument given here needs to match the argument given to `evaluate` or `integrate`, if we interpret the `unspecified` as `hessians`. Their use is slightly different but we could make things compatible, apart from the variants for the cell-centric loops maybe. I think @peterrum and I need to discuss that interface which is unfinished anyway.

kronbichler commented:
But we do not need to unify them in this PR.

kronbichler commented on tests/matrix_free/advect_1d.cc:
```diff
@@ -112,10 +112,10 @@ class MatrixFreeAdvectionBasic
         phi.read_dof_values(src);
-        phi.evaluate(true, false);
+        phi.evaluate(EvaluationFlags::values);
         for (unsigned int q = 0; q < phi.n_q_points; ++q)
           phi.submit_gradient(advection * phi.get_value(q), q);
-        phi.integrate(false, true);
+        phi.integrate(EvaluationFlags::gradients);
```
That's charming from the point of documenting itself, isn't it? :+1:

tjhei replied:
much better, yes. :-)

tjhei commented:
I have probably missed a few things, but this is ready to go from my side. (open discussion is when to deprecate things)

kronbichler commented:
May I suggest to remove the deprecation warnings on the new interfaces for now? I would like to tackle #9794 soon, and we might need to introduce some new evaluator class names for that without the degree template anyway, so this gives us a chance of getting more momentum in the interface change and to do things right there.

kronbichler commented:
To be more precise, I want to add the deprecation warnings back when we have a possibly different interface anyway so the change on the users' side is only needed once.

peterrum commented:
> To be more precise, I want to add the deprecation warnings back when we have a possibly different interface anyway so the change on the users' side is only needed once.

I agree with @kronbichler! We are planing to make some changes/additions to the `MatrixFree` interface soon. However, we have not made up our minds about the exact details...

tjhei commented:
I don't mind to remove the macro and merge without it for now. The ``@deprecated`` can stay in place and will help.

I was thinking that we could introduce a ``@deprecatedsoon`` doxygen macro that adds a @note to the function documentation. We can then regularly grep for this macro and add the deprecated define.

peterrum commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -2670,7 +2668,6 @@ class FEEvaluation : public FEEvaluationAccess<dim,
    * @deprecated Please use the gather_evaluate() function with the EvaluationFlags argument.
    */
   template <typename VectorType>
-  DEAL_II_DEPRECATED void
```
```suggestion
DEAL_II_DEPRECATED void
```
